File: src/BootCompiler/base64.pl
:- module(base64, [encode64/3,decode64/3]).

:- use_module(misc).


encode64([H,M,L|R],O,Ox) :-
  Word is (H<<16)\/(M<<8)\/L,
  encodeWord(Word,O,O1),
  encode64(R,O1,Ox).
encode64([H,M],O,Ox) :-
  W is (H<<16)\/(M<<8),
  mapToSix(W,Up,Hi,Md,_),
  encLast(Up,Hi,Md,O,['='|Ox]).
encode64([H],O,Ox) :-
  W is (H<<16),
  mapToSix(W,Up,Hi,Md,_),
  encLast(Up,Hi,Md,O,['=','='|Ox]).
encode64([],O,O).

encLast(Up,0,0,[U|O],O) :-!,
  encByte(Up,U),!.
encLast(Up,Hi,0,[U,H|O],O) :-!,
  encByte(Up,U),
  encByte(Hi,H),!.
encLast(Up,Hi,Md,[U,H,M|O],O) :-!,
  encByte(Up,U),
  encByte(Hi,H),
  encByte(Md,M),!.

mapToSix(W,Up,Hi,Md,Lw) :-
  Up is (W>>18)/\63,
  Hi is (W>>12)/\63,
  Md is (W>>6)/\63,
  Lw is (W)/\63.

encodeWord(W,[U,H,M,L|O],O) :-
  mapToSix(W,Up,Hi,Md,Lw),
  encByte(Up,U),
  encByte(Hi,H),
  encByte(Md,M),
  encByte(Lw,L).

decode64([U,H,'=','='],[B1|O],O) :-
  encByte(Up,U),
  encByte(Hi,H),
  W is (Up<<18)\/(Hi<<12),
  B1 is (W>>16)/\255.
decode64([U,H,M,'='],[B1,B2|O],O) :-
  encByte(Up,U),
  encByte(Hi,H),
  encByte(Md,M),
  W is (Up<<18)\/(Hi<<12)\/(Md<<6),
  B1 is (W>>16)/\255,
  B2 is (W>>8)/\255.
decode64([U,H,M,L|Rest],[B1,B2,B3|O],Ox) :-
  encByte(Up,U),
  encByte(Hi,H),
  encByte(Md,M),
  encByte(Lw,L),!,
  W is (Up<<18)\/(Hi<<12)\/(Md<<6)\/(Lw),
  B1 is (W>>16)/\255,
  B2 is (W>>8)/\255,
  B3 is W/\255,
  decode64(Rest,O,Ox).
decode64([],O,O).

encByte(0,'A').
encByte(1,'B').
encByte(2,'C').
encByte(3,'D').
encByte(4,'E').
encByte(5,'F').
encByte(6,'G').
encByte(7,'H').
encByte(8,'I').
encByte(9,'J').
encByte(10,'K').
encByte(11,'L').
encByte(12,'M').
encByte(13,'N').
encByte(14,'O').
encByte(15,'P').
encByte(16,'Q').
encByte(17,'R').
encByte(18,'S').
encByte(19,'T').
encByte(20,'U').
encByte(21,'V').
encByte(22,'W').
encByte(23,'X').
encByte(24,'Y').
encByte(25,'Z').
encByte(26,'a').
encByte(27,'b').
encByte(28,'c').
encByte(29,'d').
encByte(30,'e').
encByte(31,'f').
encByte(32,'g').
encByte(33,'h').
encByte(34,'i').
encByte(35,'j').
encByte(36,'k').
encByte(37,'l').
encByte(38,'m').
encByte(39,'n').
encByte(40,'o').
encByte(41,'p').
encByte(42,'q').
encByte(43,'r').
encByte(44,'s').
encByte(45,'t').
encByte(46,'u').
encByte(47,'v').
encByte(48,'w').
encByte(49,'x').
encByte(50,'y').
encByte(51,'z').
encByte(52,'0').
encByte(53,'1').
encByte(54,'2').
encByte(55,'3').
encByte(56,'4').
encByte(57,'5').
encByte(58,'6').
encByte(59,'7').
encByte(60,'8').
encByte(61,'9').
encByte(62,'+').
encByte(63,'/').

File: src/Util/base64.c
//
// Created by Francis McCabe on 2/1/17.
//

#include "base64.h"

static char codes[];

static byte decByte(byte ch);

// Decode a stream of characters as base64
retCode decode64(ioPo dst, ioPo src) {
  retCode ret = Ok;

  while (ret == Ok) {
    byte u;

    ret = inByte(src, &u);

    if (ret == Eof)           // Eof allowed on first character
      return Ok;
    else if (ret == Ok) {
      byte h;

      ret = inByte(src, &h);

      if (ret == Ok) {
        byte m;

        ret = inByte(src, &m);

        if (ret == Ok) {
          if (m == '=') {
            byte l;

            ret = inByte(src, &l);

            if (ret == Ok && l == '=') {
              byte Up = decByte(u);
              byte Hi = decByte(h);
              uint32 w = (Up << 18) | (Hi << 12);
              return outByte(dst, (byte) ((w >> 16) & 255));
            } else if (ret == Eof)
              return Eof;
            else
              return Error;
          } else {
            byte l;

            ret = inByte(src, &l);

            if (ret == Ok) {
              if (l == '=') {
                byte Up = decByte(u);
                byte Hi = decByte(h);
                byte Lo = decByte(m);
                uint32 w = (Up << 18) | (Hi << 12) | Lo << 6;
                byte b1 = (byte) ((w >> 16) & 255);
                byte b2 = (byte) ((w >> 8) & 255);

                ret = outByte(dst, b1);
                if (ret == Ok)
                  ret = outByte(dst, b2);
                return ret;
              } else {
                byte Up = decByte(u);
                byte Hi = decByte(h);
                byte Md = decByte(m);
                byte Lo = decByte(l);
                uint32 w = (Up << 18) | (Hi << 12) | Md << 6 | Lo;
                byte b1 = (byte) ((w >> 16) & 255);
                byte b2 = (byte) ((w >> 8) & 255);
                byte b3 = (byte) (w & 255);

                ret = outByte(dst, b1);
                if (ret == Ok)
                  ret = outByte(dst, b2);
                if (ret == Ok)
                  ret = outByte(dst, b3);
              }
            }
          }
        }

      }
    }
  }
  return ret;
}

static retCode encodeWord(ioPo dst, uint32 val);
static void map2Six(uint32 val, byte *up, byte *hi, byte *md, byte *lo);
static retCode encodeLast(ioPo dst, byte hi, byte md, byte lo, char *rest);
static retCode encByte(ioPo dst, byte b);

// Encode a stream of bytes into base64 characters
retCode encode64(ioPo src, ioPo dst) {
  retCode ret = Ok;

  while (ret == Ok) {
    byte h;
    byte Up, Hi, Md, Lo;

    ret = inByte(src, &h);

    if (ret == Ok) {
      byte m;

      ret = inByte(src, &m);

      if (ret == Ok) {
        byte l;
        ret = inByte(src, &l);

        if (ret == Ok) {
          ret = encodeWord(dst, (((unsigned int) h & 0xff) << 16) | ((m & 0xff) << 8) | (l & 0xff));
        } else if (ret == Eof) {
          map2Six((((unsigned int) h & 0xff) << 16) | ((m & 0xff) << 8), &Up, &Hi, &Md, &Lo);
          return encodeLast(dst, Up, Hi, Md, "=");
        }
      } else if (ret == Eof) {
        map2Six((((unsigned int) h & 0xff) << 16), &Up, &Hi, &Md, &Lo);
        return encodeLast(dst, Up, Hi, Md, "==");
      }
    } else
      return ret;
  }
  return ret;
}

static void map2Six(uint32 val, byte *up, byte *hi, byte *md, byte *lo) {
  *up = (byte) ((val >> 18) & 63);
  *hi = (byte) ((val >> 12) & 63);
  *md = (byte) ((val >> 6) & 63);
  *lo = (byte) (val & 63);
}

static retCode encodeWord(ioPo dst, uint32 val) {
  byte Up, Hi, Md, Lo;

  map2Six(val, &Up, &Hi, &Md, &Lo);

  retCode ret = encByte(dst, Up);
  if (ret == Ok)
    ret = encByte(dst, Hi);
  if (ret == Ok)
    ret = encByte(dst, Md);
  if (ret == Ok)
    ret = encByte(dst, Lo);
  return ret;
}

static retCode encodeLast(ioPo dst, byte hi, byte md, byte lo, char *rest) {
  retCode ret = encByte(dst, hi);
  if (ret == Ok)
    ret = encByte(dst, md);
  if (ret == Ok && lo != 0)
    ret = encByte(dst, lo);
  if (ret == Ok)
    ret = outStr(dst, rest);
  return ret;
}

static retCode encByte(ioPo dst, byte b) {
  return outByte(dst, (byte) codes[b]);
}

// Core table for encoding
static char codes[] = {
  'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W',
  'X', 'Y', 'Z',
  'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w',
  'x', 'y', 'z',
  '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '+', '/'
};

static byte decByte(byte code) {
  for (int ix = 0; ix < NumberOf(codes); ix++) {
    if (codes[ix] == code)
      return (byte) ix;
  }
  return (byte) -1;
}
